sdk/data/aztables/internal/auth/challenge_policy.go

```diff
@@ -0,0 +1,236 @@
+	as := acquiringResourceState{
+		p:   scp,
+		req: req,
+	}
+
+	if scp.scope == nil || scp.tenantID == nil {
+		// First request, get both to get the token
+		challengeReq, err := scp.getChallengeRequest(*req)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := challengeReq.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode > 399 && resp.StatusCode != http.StatusUnauthorized {
+			// the request failed for some other reason, don't try any further
+			return resp, nil
+		}
+		err = scp.findScopeAndTenant(resp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tk, err := scp.mainResource.GetResource(as)
+	if err != nil {
+		return nil, err
+	}
+
+	if token, ok := tk.(*azcore.AccessToken); ok {
+		req.Raw().Header.Set(
+			headerAuthorization,
+			fmt.Sprintf("%s%s", bearerHeader, token.Token),
+		)
+	}
+
+	// send a copy of the request
+	cloneReq := req.Clone(req.Raw().Context())
+	resp, cloneReqErr := cloneReq.Next()
+	if cloneReqErr != nil {
+		return nil, cloneReqErr
+	}
+
+	// If it fails and has a 401, try it with a new token
+	if resp.StatusCode == 401 {
+		// Force a new token
+		scp.mainResource.Reset()
+
+		// Find the scope and tenant again in case they have changed
+		err := scp.findScopeAndTenant(resp)
+		if err != nil {
+			// Error parsing challenge, doomed to fail. Return
+			return resp, cloneReqErr
+		}
+
+		tk, err := scp.mainResource.GetResource(as)
+		if err != nil {
+			return resp, err
+		}
+
+		if token, ok := tk.(*azcore.AccessToken); ok {
+			req.Raw().Header.Set(
+				headerAuthorization,
+				bearerHeader+token.Token,
+			)
+		} else {
+			// tk is not an azcore.AccessToken type, something went wrong and we should return the 401 and accompanying error
+			return resp, cloneReqErr
+		}
+
+		// send the original request now
+		return req.Next()
+	}
+
+	return resp, err
+}
+
+// parses Tenant ID from auth challenge
+// https://login.microsoftonline.com/00000000-0000-0000-0000-000000000000
+func parseTenant(url string) *string {
+	if url == "" {
+		return to.Ptr("")
+	}
+	parts := strings.Split(url, "/")
+	tenant := parts[3]
+	tenant = strings.ReplaceAll(tenant, ",", "")
+	return &tenant
+}
+
+type challengePolicyError struct {
+	err error
+}
+
+func (c *challengePolicyError) Error() string {
+	return c.err.Error()
+}
+
+func (*challengePolicyError) NonRetriable() {
+	// marker method
+}
+
+func (c *challengePolicyError) Unwrap() error {
+	return c.err
+}
+
+var _ errorinfo.NonRetriable = (*challengePolicyError)(nil)
+
+// sets the scp.scope and scp.tenantID from the WWW-Authenticate header
+func (scp *TablesChallengePolicy) findScopeAndTenant(resp *http.Response) error {
+	authHeader := resp.Header.Get("WWW-Authenticate")
+	if authHeader == "" {
+		return &challengePolicyError{err: errors.New("response has no WWW-Authenticate header for challenge authentication")}
+	}
+
+	// Strip down to auth and resource
+	// Format is "Bearer authorization=\"<site>\" resource=\"<site>\"" OR
```

Comment:
I recall Storage challenges as having unquoted values, are these from another RP?

Reply:
I looked at the response, you are correct. I have updated this